### PR TITLE
[precompiled] Initial proof-of-concept with App.cpp

### PR DESCRIPTION
**Pull request by dherrera-fb**
This is a proof-of-concept PR to show a potential implementation of https://github.com/CLIUtils/CLI11/issues/338

- Add C11_COMPILE cmake option that creates a static lib instead of header-only
- Add C11_INLINE macro that depends on C11_COMPILE
- Split App.hpp into App.hpp and impl/App_inl.hpp
- Add App.cpp that compiles App_inl.hpp into an object file
- CMake modifications to handle impl headers differently for sinlge-header, headers-only, and compiled versions

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -160,13 +166,35 @@ if(NOT CMAKE_VERSION VERSION_LESS 3.13)
...
+  list(INSERT MAYBE_CONFIGURE_DEPENDS 0 CONFIGURE_DEPENDS)
+endif()
+
+file(GLOB CLI11_headers ${MAYBE_CONFIGURE_DEPENDS} "${PROJECT_SOURCE_DIR}/include/CLI/*.hpp")
+file(GLOB CLI11_impl_headers ${MAYBE_CONFIGURE_DEPENDS}
+     "${PROJECT_SOURCE_DIR}/include/CLI/impl/*.hpp")
+
+if(CLI11_PRECOMPILED)
+  # Create static lib
+  file(GLOB CLI11_precompile_sources "${PROJECT_SOURCE_DIR}/src/*.cpp")
```
Comment (jzakrzewski) [inline CMakeLists.txt:181]: Do you really think that globbing source files is a good idea?

Comment (dherrera-fb) [reply CMakeLists.txt:181]: I have no opinion on globbing. Master branch is using glob so this PR is following the status-quo. I'm happy to update it if glob is removed from the main branch.

Comment (jzakrzewski) [reply CMakeLists.txt:181]: Yeah - it's already bad in master but this makes it even worse when the library will get build with outdated source list during development it's time wasted for chasing non-existent problems.
I know some people do not want to list all the files in the build scripts but I'd urge the maintainers to consider that.

Comment (henryiii) [reply CMakeLists.txt:181]: `CONFIGURE_DEPENDS` makes it quite a bit better, and we add that if CMake supports it.

Comment (henryiii) [reply CMakeLists.txt:181]: (though I think it's missing here, actually)

**Pull request closed without merging**